src/main/java/com/google/sps/Scheduler.java

```diff
@@ -1,15 +1,134 @@
+      if (course.isRequired()) {
+        requiredCourses.add(course);
+      } else {
+        nonRequiredCourses.add(course);
+      }
+    }
+
+    // Generate all possible valid Schedules that include only required courses.
+    float reqCredits = totalCredits(requiredCourses);
+    Set<List<Course>> requiredSet = new HashSet<>();
+
+    generateSchedulesHelper(requiredCourses, new ArrayList<>(), 
+        new Invariants(reqCredits, reqCredits), requiredSet);
+    
+    // Using those Schedules as a building block, generate all valid schedules
+    Set<List<Course>> scheduleSet = new HashSet<>();
+    for (List<Course> validCourseList : requiredSet) {
+      generateSchedulesHelper(nonRequiredCourses, validCourseList, invariants, scheduleSet);
+    }
+
+    List<Schedule> schedules = new ArrayList<>();
+    for (List<Course> courseList: scheduleSet) {
+      schedules.add(new Schedule(courseList));
+    }
+
+    return schedules;
+  }
+
+  /**
+   * Recursively iterates through a list of courses and adds schedules to the
+   * given set.
+   * 
+   * @param courses     The list of courses to iterate through
+   * @param courseList  The current course list to be added/operated on
+   * @param invariants  The invariants that constrain the valid schedules
+   * @param courseLists The set of course lists that is being generated/added to.
+   */
+  private void generateSchedulesHelper(List<Course> courses, List<Course> courseList, 
+      Invariants invariants, Set<List<Course>> courseLists) {
+
+    // This means that we have a valid schedule, and should add it to the list.
+    float currCredits = totalCredits(courseList);
+    if (invariants.meetsCreditRequirement(currCredits)) {
+      courseLists.add(courseList);
+    }
+
+    // We've reached the end of the list or have a full schedule, we are finished.
+    if (courses.isEmpty() || currCredits >= invariants.getMaxCredits()) {
+      return;
+    }
+
+    Course course = courses.remove(0);
+    List<Section> courseSections = course.getSections();
+    List<Section> scheduleSections = courseList.stream()
+        .map(c -> c.getSections().get(0))
+        .collect(Collectors.toList());
+
+    /*
+     * Iterate over each section in the current course and check for overlap. If the
+     * section doesn't overlap and adding it would result in a valid schedule, add
+     * it and make a new recursive call.
+     */
+    // TODO: Implement logic for required courses having no valid schedules
+    for (Section section : courseSections) {
+      boolean doesNotOverlapSchedule = !sectionsOverlap(section, scheduleSections);
+      boolean doesNotExceedCreditLimit = 
+          currCredits + course.getCredits() <= invariants.getMaxCredits();
+      
+      if (doesNotOverlapSchedule && doesNotExceedCreditLimit) {
+        List<Course> newCourseList = new ArrayList<>(courseList);
+        newCourseList.add(new Course(course.getName(), course.getCourseID(), course.getSubject(),
```

Comment:
It seems like you're using the `Course` class in two slightly different ways:
1. To represent a course offering with multiple available sections
2. To represent a specific course section a student is enrolled in

It would be best for these two things to either be well represented with separate data structures, or represented by data structures that compose differently to represent both.

One approach that requires a less substantial change could be to use a different structure to represent course enrollment:

```
class CourseEnrollment {
  public Course getCourse();
  public Section getEnrolledSection();
}
```

It's not necessarily the cleanest solution here, but it works well enough and avoids this mild misuse of the `Course` class.

Reply:
We're creating a new class for courses that have been scheduled (called ScheduledCourse). It's essentially the same as the original Course class, but the `sections` field is now a single `section` as opposed to a list.